judgels-backends/uriel/uriel-app/src/test/java/judgels/uriel/contest/scoreboard/troc/TrocScoreboardProcessorTests.java

```diff
@@ -0,0 +1,780 @@
+package judgels.uriel.contest.scoreboard.troc;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import judgels.gabriel.api.Verdict;
+import judgels.jophiel.api.profile.Profile;
+import judgels.sandalphon.api.submission.programming.Submission;
+import judgels.uriel.api.contest.Contest;
+import judgels.uriel.api.contest.ContestStyle;
+import judgels.uriel.api.contest.contestant.ContestContestant;
+import judgels.uriel.api.contest.module.StyleModuleConfig;
+import judgels.uriel.api.contest.module.TrocStyleModuleConfig;
+import judgels.uriel.api.contest.scoreboard.ScoreboardState;
+import judgels.uriel.api.contest.scoreboard.TrocScoreboard.TrocScoreboardEntry;
+import judgels.uriel.api.contest.scoreboard.TrocScoreboard.TrocScoreboardProblemState;
+import judgels.uriel.contest.scoreboard.AbstractProgrammingScoreboardProcessorTests;
+import judgels.uriel.contest.scoreboard.ScoreboardProcessResult;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class TrocScoreboardProcessorTests extends AbstractProgrammingScoreboardProcessorTests {
+    private TrocScoreboardProcessor scoreboardProcessor = new TrocScoreboardProcessor();
+
+    @Nested
+    class Process {
+        private ScoreboardState state = new ScoreboardState.Builder()
+                .addProblemJids("p1", "p2")
+                .addProblemAliases("A", "B")
+                .problemPoints(ImmutableList.of(1, 10))
+                .build();
+
+        private Contest contest = new Contest.Builder()
+                .beginTime(Instant.ofEpochSecond(60))
+                .duration(Duration.ofMinutes(100))
+                .id(1)
+                .jid("JIDC")
+                .name("contest-name")
+                .slug("contest-slug")
+                .style(ContestStyle.TROC)
+                .build();
+
+        private StyleModuleConfig styleModuleConfig = new TrocStyleModuleConfig.Builder()
+                .wrongSubmissionPenalty(10)
+                .build();
+
+        private Set<ContestContestant> contestants = ImmutableSet.of(
+                new ContestContestant.Builder().userJid("c1").build(),
+                new ContestContestant.Builder().userJid("c2").contestStartTime(Instant.ofEpochSecond(300)).build());
+
+        private Map<String, Profile> profilesMap = ImmutableMap.of(
+                "c1", new Profile.Builder().username("c1").build(),
+                "c2", new Profile.Builder().username("c2").build());
+
+        @Test
+        void time_calculation() {
```

Comment:
Can we add some CE case and WA after AC case here? (Making sure that they don't be counted in time/penalty calculation)